realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -600,14 +640,275 @@ public static synchronized void migrateRealm(RealmConfiguration configuration, R
+
+                                QueryUpdateTask queryUpdateTask = QueryUpdateTask.newBuilder()
+                                        .realmConfiguration(getConfiguration())
+                                        .addObject(realmObjectWeakReference,
+                                                realmQuery.handoverQueryPointer(),
+                                                realmQuery.getArgument())
+                                        .sendToHandler(handler, REALM_COMPLETED_ASYNC_FIND_FIRST)
+                                        .build();
+
+                                asyncQueryExecutor.submit(queryUpdateTask);
+                            } else {
+                                // should not happen, since the the background thread position itself against the provided version
+                                // and the caller thread can only go forward (advance_read)
+                                throw new IllegalStateException("Caller thread behind the worker thread");
+                            }
+                        } // else: element GC'd in the meanwhile
+                    }
+                    break;
+                }
             }
             return true;
         }
     }
 
+    /**
+     * This will prevent advanceReading from accidentally advancing the thread and potentially re-run the queries in this thread.
+     * @return {@code true} if there is at least one (non GC'd) instance of {@link RealmResults} {@code false} otherwise
+     */
+    private boolean threadContainsAsyncQueries () {
+        deleteWeakReferences();
+        boolean isEmpty = true;
+        Iterator<Map.Entry<WeakReference<RealmResults<? extends RealmObject>>, RealmQuery<?>>> iterator = asyncRealmResults.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<WeakReference<RealmResults<? extends RealmObject>>, RealmQuery<?>> next = iterator.next();
+            if (next.getKey().get() == null) {
+                // clean the GC'ed instances
+                // we could've avoided this if we had a 'WeakIdentityHashmap' data structure. miss Guava :(
+                iterator.remove();
+            } else {
+                isEmpty = false;
+            }
+        }
+
+        return !isEmpty;
+    }
+
+    private void deleteWeakReferences() {
+        // From the AOSP FinalizationTest:
+        // https://android.googlesource.com/platform/libcore/+/master/support/src/test/java/libcore/
+        // java/lang/ref/FinalizationTester.java
+        // System.gc() does not garbage collect every time. Runtime.gc() is
+        // more likely to perform a gc.
+        Runtime.getRuntime().gc();
+        Reference<? extends RealmResults<? extends RealmObject>> weakReference;
+        while ((weakReference = referenceQueue.poll()) != null ) { // Does not wait for a reference to become available.
+            RealmLog.d("deleted 1 reference: " + weakReference);
```

Comment:
We do really wan't to fill the log with these? Even though the Logger might not output anything you will still have the overhead of methods calls and boolean checks + You are creating a StringBuilder here.